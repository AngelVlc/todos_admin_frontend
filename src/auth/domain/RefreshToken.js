export class RefreshToken {
  constructor({id, userId, expirationDate}) {
    this._id = id;
    this._userId = userId;
    this._expirationDate = expirationDate;
    this._selected = false;
  }

  get id() {
    return this._id;
  }

  get userId() {
    return this._userId;
  }

  get expirationDate() {
    return this._expirationDate;
  }

  get selected() {
    return this._selected;
  }

  set selected(value) {
    this._selected = value;
  } 
}