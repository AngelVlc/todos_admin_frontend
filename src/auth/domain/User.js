export class User {
  constructor({ id, name, isAdmin }) {
    this.id = id;
    this.name = name;
    this.isAdmin = isAdmin;
    this.password = "";
    this.confirmPassword = "";
  }

  static createEmpty() {
    return new User({ name: "", isAdmin: false });
  }
}
