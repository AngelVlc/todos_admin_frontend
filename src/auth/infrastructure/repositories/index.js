export * from './RefreshTokensRepository';
export * from './UsersRepository';
