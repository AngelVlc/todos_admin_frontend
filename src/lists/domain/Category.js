export class Category {
  constructor({ id, name, description }) {
    this.id = id;
    this.name = name;
    this.description = description;
  }

  static createEmpty() {
    return new Category({ id: -1, name: "", description: "" });
  }
}
