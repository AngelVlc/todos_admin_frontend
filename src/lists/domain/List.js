export class List {
  constructor({ id, name, itemsCount, items = [] }) {
    this.id = id;
    this.name = name;
    this.items = items;
    this.itemsCount = itemsCount;
  }

  static createEmpty() {
    return new List({ id: -1, name: "", itemsCount: 0 });
  }

  addNewItem(newItem) {
    newItem.id = -1;

    for (const item of this.items) {
      if (item.id <= newItem.id) {
        newItem.id = item.id - 1;
      }
    };

    this.items.push(newItem);
  }
}
