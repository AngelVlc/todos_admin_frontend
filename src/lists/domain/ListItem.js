export class ListItem {
  constructor({ id, title, description, listId }) {
    this.id = id;
    this.title = title;
    this.description = description;
    this.listId = listId;
  }

  static createEmpty(listId) {
    return new ListItem({ id: -1, listId: listId, title: "", description: "" });
  }
}
