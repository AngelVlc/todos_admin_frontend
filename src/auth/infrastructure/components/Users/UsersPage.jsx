import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link, useHistory } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { GetUsersUseCase } from "../../../application/users";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb";

export const UsersPage = () => {
  const [users, setUsers] = useState([]);
  const { useCaseFactory } = useContext(AppContext);
  let history = useHistory();

  const getUsers = useCallback(async () => {
    const getUsersUseCase = useCaseFactory.get(GetUsersUseCase);
    const data = await getUsersUseCase.execute();

    setUsers(data);
  }, [useCaseFactory]);

  useEffect(() => {
    getUsers();
  }, [getUsers]);

  return (
    <div className="container">
      <Breadcrumb
        items={[
          { url: "/users", text: "Users" },
        ]}
      />
      <h3 className="title">USERS</h3>
      <table className="table">
        <thead>
          <tr>
            <td>User</td>
            <td>Admin?</td>
            <td>
              <button
                className="button is-small"
                data-testid="addNew"
                onClick={() => history.push("users/new")}
              >
                <span className="icon is-small">
                  <i className="fas fa-plus"></i>
                </span>
              </button>
            </td>
          </tr>
        </thead>
        <tbody>
          {users.length > 0 &&
            users.map((user) => (
              <tr key={user.id}>
                <td>
                  <Link
                    className="has-text-black"
                    data-testid={`editUser${user.id}`}
                    to={`/users/${user.id}/edit`}
                  >
                    {user.name}
                  </Link>
                </td>
                <td>
                  {user.isAdmin && (
                    <center>
                      <span className="icon is-small">
                        <i className="fas fa-check fa-xs"></i>
                      </span>
                    </center>
                  )}
                </td>
                <td>
                  <center>
                    <Link
                      className="has-text-black delete"
                      data-testid={`deleteUser${user.id}`}
                      to={`/users/${user.id}/delete`}
                    >
                    </Link>
                  </center>
                </td>
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  );
};
