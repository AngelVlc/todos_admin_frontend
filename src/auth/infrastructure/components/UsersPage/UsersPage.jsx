import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link, useHistory } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { GetUsersUseCase } from "../../../application/users";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb";
import "./UsersPage.css";

export const UsersPage = () => {
  const [users, setUsers] = useState();
  const { useCaseFactory } = useContext(AppContext);
  let history = useHistory();

  const getUsers = useCallback(async () => {
    const getUsersUseCase = useCaseFactory.get(GetUsersUseCase);
    const data = await getUsersUseCase.execute();

    setUsers(data);
  }, [useCaseFactory]);

  useEffect(() => {
    getUsers();
  }, [getUsers]);

  return (
    <>
      {users && (
        <div className="container">
          <Breadcrumb items={[{ url: "/users", text: "Users" }]} />
          <h3 className="title">USERS</h3>
          <table className="table">
            <thead>
              <tr>
                <td>User</td>
                <td>Admin?</td>
                <td>
                  <button
                    className="button is-small"
                    data-testid="addNew"
                    onClick={() => history.push("users/new")}
                  >
                    <span className="icon is-small">
                      <i className="fas fa-plus"></i>
                    </span>
                  </button>
                </td>
              </tr>
            </thead>
            <tbody>
              {users.length > 0 &&
                users.map((user) => (
                  <tr key={user.id}>
                    <td>
                      <Link
                        className="has-text-black td-content"
                        data-testid={`editUser${user.id}`}
                        to={`/users/${user.id}/edit`}
                      >
                        {user.name}
                      </Link>
                    </td>
                    <td>
                      <Link
                        className="has-text-black td-content"
                        to={`/users/${user.id}/edit`}
                      >
                        {user.isAdmin && (
                          <center>
                            <span className="icon is-small">
                              <i className="fas fa-check fa-xs"></i>
                            </span>
                          </center>
                        )}
                      </Link>
                    </td>
                    <td>
                      <center>
                        <Link
                          className="has-text-black delete"
                          data-testid={`deleteUser${user.id}`}
                          to={`/users/${user.id}/delete`}
                        ></Link>
                      </center>
                    </td>
                  </tr>
                ))}
            </tbody>
          </table>
        </div>
      )}
    </>
  );
};
