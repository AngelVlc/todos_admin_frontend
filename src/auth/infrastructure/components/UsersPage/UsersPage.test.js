import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { UsersPage } from "./UsersPage";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { User } from "../../../domain";

afterEach(cleanup);

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const history = createMemoryHistory();

const renderWithContextAndRouter = () => {
  const fakeGetUsersUseCase = {
    execute: () => [
      new User({ id: 1, name: "user1", isAdmin: true }),
      new User({ id: 2, name: "user2", isAdmin: false }),
    ],
  };

  const useCaseFactory = {
    get: () => fakeGetUsersUseCase,
  };

  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <Router history={history}>{<UsersPage />}</Router>
      </AppContext.Provider>
    ),
  };
};

describe("UsersPage", () => {
  it("should match the snapshot", async () => {
    let fragment;
    await act(async () => {
      const { asFragment } = renderWithContextAndRouter();
      fragment = asFragment;
    });

    expect(fragment()).toMatchSnapshot();
  });

  it("should add a new user", async () => {
    let container;
    await act(async () => {
      container = renderWithContextAndRouter();
    });

    await waitFor(() => {
      fireEvent.click(container.getByTestId("addNew"));
    });

    expect(mockHistoryPush).toHaveBeenCalled();
    expect(mockHistoryPush.mock.calls[0][0]).toBe("users/new");
    mockHistoryPush.mockClear();
  });

  it("should edit the user", async () => {
    let container;
    await act(async () => {
      container = renderWithContextAndRouter();
    });

    await waitFor(() => {
      fireEvent.click(container.getByTestId("editUser2"));
    });

    expect(mockHistoryPush).toHaveBeenCalled();
    expect(mockHistoryPush.mock.calls[0][0]).toBe("/users/2/edit");
    mockHistoryPush.mockClear();
  });

  it("should delete the user", async () => {
    let container;
    await act(async () => {
      container = renderWithContextAndRouter();
    });

    await waitFor(() => {
      fireEvent.click(container.getByTestId("deleteUser2"));
    });

    expect(history.location.pathname).toBe("/users/2/delete");
  });
});
