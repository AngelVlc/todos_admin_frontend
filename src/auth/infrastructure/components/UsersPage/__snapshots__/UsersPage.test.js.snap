// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersPage should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/users"
          >
            Users
          </a>
        </li>
      </ul>
    </nav>
    <h3
      class="title"
    >
      USERS
    </h3>
    <table
      class="table"
    >
      <thead>
        <tr>
          <td>
            User
          </td>
          <td>
            Admin?
          </td>
          <td>
            <button
              class="button is-small"
              data-testid="addNew"
            >
              <span
                class="icon is-small"
              >
                <i
                  class="fas fa-plus"
                />
              </span>
            </button>
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a
              class="has-text-black td-content"
              data-testid="editUser1"
              href="/users/1/edit"
            >
              user1
            </a>
          </td>
          <td>
            <a
              class="has-text-black td-content"
              href="/users/1/edit"
            >
              <center>
                <span
                  class="icon is-small"
                >
                  <i
                    class="fas fa-check fa-xs"
                  />
                </span>
              </center>
            </a>
          </td>
          <td>
            <center>
              <a
                class="has-text-black delete"
                data-testid="deleteUser1"
                href="/users/1/delete"
              />
            </center>
          </td>
        </tr>
        <tr>
          <td>
            <a
              class="has-text-black td-content"
              data-testid="editUser2"
              href="/users/2/edit"
            >
              user2
            </a>
          </td>
          <td>
            <a
              class="has-text-black td-content"
              href="/users/2/edit"
            />
          </td>
          <td>
            <center>
              <a
                class="has-text-black delete"
                data-testid="deleteUser2"
                href="/users/2/delete"
              />
            </center>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
